plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java

```diff
@@ -202,6 +212,73 @@ private static boolean isAggregate(ExpressionContext expressionContext)
         return expressionContext.getType() == ExpressionContext.Type.FUNCTION && expressionContext.getFunction().getType() == FunctionContext.Type.AGGREGATION;
     }
 
+    private static boolean hasAggregate(ExpressionContext expressionContext)
+    {
+        switch (expressionContext.getType()) {
+            case IDENTIFIER:
+            case LITERAL:
+                return false;
+            case FUNCTION:
+                if (isAggregate(expressionContext)) {
+                    return true;
+                }
+                for (ExpressionContext argument : expressionContext.getFunction().getArguments()) {
+                    if (hasAggregate(argument)) {
+                        return true;
+                    }
+                }
+                return false;
+        }
+        throw new PinotException(PINOT_EXCEPTION, Optional.empty(), format("Unsupported expression type '%s'", expressionContext.getType()));
+    }
+
+    // Extracted from org.apache.pinot.core.query.reduce.AggregationDataTableReducer
+    private static DataSchema getPreAggregationDataSchema(QueryContext queryContext)
+    {
+        AggregationFunction[] aggregationFunctions = queryContext.getAggregationFunctions();
+        int numAggregationFunctions = aggregationFunctions.length;
+        String[] columnNames = new String[numAggregationFunctions];
+        DataSchema.ColumnDataType[] columnDataTypes = new DataSchema.ColumnDataType[numAggregationFunctions];
+        for (int i = 0; i < numAggregationFunctions; i++) {
+            AggregationFunction aggregationFunction = aggregationFunctions[i];
+            columnNames[i] = aggregationFunction.getResultColumnName();
+            columnDataTypes[i] = aggregationFunction.getFinalResultColumnType();
+        }
+        return new DataSchema(columnNames, columnDataTypes);
+    }
+
+    private static Map<String, PinotColumnNameAndTrinoType> getAggregateTypes(SchemaTableName schemaTableName, QueryContext queryContext, Map<String, ColumnHandle> columnHandles)
+    {
+        // A mapping from pinot expression to the returned pinot column name and trino type
+        // Note: the column name is set by the postaggregation schema
```

Comment:
postaggregation schema -> PostAggregationHandler.

IMO it's easier to refer to where to look that way.